letsencrypt/client/revoker.py

```diff
@@ -6,146 +14,513 @@
 from letsencrypt.acme import util as acme_util
 
-from letsencrypt.client import crypto_util
-from letsencrypt.client import display
-from letsencrypt.client import interfaces
+from letsencrypt.client import errors
+from letsencrypt.client import le_util
 from letsencrypt.client import network
 
+from letsencrypt.client.display import util as display_util
+from letsencrypt.client.display import revocation
+
 
 class Revoker(object):
     """A revocation class for LE.
 
+    .. todo:: Add a method to specify your own certificate for revocation - CLI
+
+    :ivar network: Network object
+    :type network: :class:`letsencrypt.client.network`
+
+    :ivar installer: Installer object
+    :type installer: :class:`~letsencrypt.client.interfaces.IInstaller`
+
     :ivar config: Configuration.
-    :type config: :class:`letsencrypt.client.interfaces.IConfig`
+    :type config: :class:`~letsencrypt.client.interfaces.IConfig`
 
-    """
+    :ivar bool no_confirm: Whether or not to ask for confirmation for revocation
 
-    def __init__(self, installer, config):
+    """
+    def __init__(self, installer, config, no_confirm=False):
         self.network = network.Network(config.server)
         self.installer = installer
         self.config = config
+        self.no_confirm = no_confirm
+
+        le_util.make_or_verify_dir(config.cert_key_backup, 0o700, os.geteuid())
 
-    def acme_revocation(self, cert):
-        """Handle ACME "revocation" phase.
+        # TODO: Find a better solution for this...
+        self.list_path = os.path.join(config.cert_key_backup, "LIST")
+        # Make sure that the file is available for use for rest of class
+        open(self.list_path, "a").close()
 
-        :param dict cert: TODO
+    def revoke_from_key(self, authkey):
+        """Revoke all certificates under an authorized key.
 
-        :returns: ACME "revocation" message.
-        :rtype: :class:`letsencrypt.acme.message.Revocation`
+        :param authkey: Authorized key used in previous transactions
+        :type authkey: :class:`letsencrypt.client.le_util.Key`
 
         """
-        certificate = acme_util.ComparableX509(
-            M2Crypto.X509.load_cert(cert["backup_cert_file"]))
-        with open(cert["backup_key_file"], 'rU') as backup_key_file:
-            key = Crypto.PublicKey.RSA.importKey(backup_key_file.read())
+        certs = []
+        clean_pem = Crypto.PublicKey.RSA.importKey(authkey.pem).exportKey("PEM")
+        with open(self.list_path, "rb") as csvfile:
+            csvreader = csv.reader(csvfile)
+            for row in csvreader:
+                # idx, cert, key
+                # Add all keys that match to marked list
+                # Note: The key can be different than the pub key found in the
+                #    certificate.
+                _, b_k = self._row_to_backup(row)
+                if clean_pem == Crypto.PublicKey.RSA.importKey(
+                        open(b_k).read()).exportKey("PEM"):
```

Comment:
Given the assumptions of the module... keys_certs are only added after being obtained from the ACME client (if there are invalid keys then we should have failed long ago... and LIST is treated as the master key.  (There are never any files listed that aren't already in the appropriate place.) I don't believe the CSV input should be a problem.... but while reviewing this I did notice line 67 is a problem.  Thanks.